Issue #961 elasticsearch index

kess-net opened the pull request:
This should solve https://github.com/RotherOSS/otobo/issues/961. Please review.

kess-net committed:
issue #961 elasticsearch index

Introduce separeate settings for each elasticserach index.

kess-net committed:
issue #961 elasticsearch_index

Fix Migration.pm and correct test accordinly.
Also use Test2:V0 directly

bschmalhofer reviewed (commented):
The ` Config   => $ConfigIndexSettings && $ConfigIndexSettings{Ticket} || $Config, ` looks a bit unusual  to me. It's a neat trick but I would add some parens to make obvious what is going on.
` Config   => (($ConfigIndexSettings && $ConfigIndexSettings{Ticket}) || $Config),`

bschmalhofer commented on scripts/test/Elasticsearch/IndexSetting.t:
```diff
@@ -35,48 +37,57 @@ my $MigObject = $Kernel::OM->Get('Kernel::System::Console::Command::Maint::Elast
-$Self->Is(
-    $Index1->{'index.mapping.total_fields.limit'}, 1000,
-    '#1 index.mapping.total_fields.limit');
+Test2::V0::is(
```
In test scripts the best practice is to use simply `is()`. E.g.
`is( $index1, { NS => 5, NR => 10, FieldsLimit => 1000 }, 'index1' );`

Besides, I'm confused again. I thought that `_IndexSettingsGet()` should transform the Config to a data structure with keys like `number_of_shards`.

kess-net replied:
That depends on the config settings. As you can see in the expansion tests below in this file it depends on the layout which is set with 'Elasticsearch::IndexSettings'. The first four tests do not use a verbose layout.

kess-net committed:
issue #961 elasticsearch_index

minor readability changes

svenoe commented on Kernel/Config/Files/XML/Framework.xml:
```diff
@@ -9217,17 +9217,59 @@ via the Preferences button after logging in.
-         If not set the general settings defined al 'all' are used. -->
-    <Setting Name="Elasticsearch::ArticleIndexCreationSettings###Ticket" Required="0" Valid="1">
-        <Description Translatable="1">Number of shards (NS) and replicas (NR) for the index ticket.</Description>
+    <Setting Name="Elasticsearch::IndexSettings" Required="0" Valid="0">
```
Be careful here. This is technically not wrong, but I would prefer, if you used Elasticsearch::IndexSettings###Default. At the moment $ConfigObject->Get('Elasticsearch::IndexSettings') returns the perl structure:
`NS => 1,
NR => 0,
FieldsLimit => 2000,
Ticket => {
    NS => 1,
    ...,
},
Customer => {...},
...,`
While this can be used, as we want people to be flexible with adding more stuff to the index configuration and thus maybe ConfigOptions themselves, too, I would rather have a strict separation there.

kess-net replied:
ok, point taken.

svenoe commented on Kernel/Config/Files/XML/Framework.xml:
```diff
@@ -9206,8 +9206,8 @@ via the Preferences button after logging in.
     </Setting>
-    <Setting Name="Elasticsearch::ArticleIndexCreationSettings###all" Required="0" Valid="1">
-        <Description Translatable="1">Number of shards (NS) and replicas (NR) for the index.</Description>
+    <Setting Name="Elasticsearch::ArticleIndexCreationSettings" Required="0" Valid="1">
```
Maybe I overlooked this somewhere, but could you please look up, what the FieldsLimit actually does, and whether we really want to add it, please? (This is an addition of Tomohisa and would be new with this patch, too.)

kess-net replied:
Personally I don't think we need FieldsLimit by default. This defines the maximum number of fields in an index. The default setting is 1000. It is meant to prevent very large mappings and therefore poorly performing searches.
I did understand Tomohisas introduction here as an example for what is possible.
Maybe we should remove it from our configuration and instead deliver an explained example for a more extended configuration.

kess-net replied:
Forget my previous comment. I decided to make usage of the FieldsLimit setting in the installer. I think, Tomohisa intended do this but did not finished it for whatever reason.

kess-net replied:
Forget my previous comment. I decided to make usage of the FieldsLimit setting in the installer. I think, Tomohisa intended do this but did not finished it for whatever reason.

svenoe replied:
Is there anything speaking against setting it to 1000, if it's the default? Probably it doesn't matter much, but it's just keeping everything exactly the same for people who are currently using it.

svenoe commented on Kernel/System/Console/Command/Maint/Elasticsearch/Migration.pm:
```diff
@@ -129,32 +135,32 @@ sub Run {
         $Self->MigrateCompanies(
             ESObject => $ESObject,
-            Config   => $Config->{Customer} || $Config->{all},
+            Config   => ( ($ConfigIndexSettings && $ConfigIndexSettings->{Customer}) || $Config ),
```
I'm a bit torn on this, honestly. I mean on the one hand it's nice, on the other hand, I know that some people will stumble upon this...for now it can stay, I guess, especially as with 10.1 we will sort our the old configuration anyways.

kess-net replied:
Well,... I like compressed code rather than long and verbose blocks of lines reading more like Basic. Sorry if this is really too difficult. I'll try to use a more simple coding in the future.

kess-net replied:
Maybe I should have used the tenary operator?

bschmalhofer replied:
It does look a bit unusual. I probably would have written the check as:
```
my $ConfigIndexSettings = $Kernel::OM->Get('Kernel::Config')->Get('Elasticsearch::IndexSettings') // {};
...
Config   => ($ConfigIndexSettings->{Customer}) // $Config ),
```

kess-net replied:
Ah, yes!  Thank you. I forgot about this operator.

svenoe commented on Kernel/System/Console/Command/Maint/Elasticsearch/Migration.pm:
```diff
@@ -690,6 +696,9 @@ sub _IndexSettingsGet {
     }
 
     my $IndexSettingsTemplate = $Kernel::OM->Get('Kernel::Config')->Get('Elasticsearch::IndexSettings');
```
This is wrong as it is. Generally there is a problem with the naming I was not aware of until now. You are using Elasticsearch::IndexSettings as the new ElasticsearchArticleIndexCreationSettings, which I agree is the better name, as we use it for various indices. In the initial pull request of Tomohisa though, the same name was used for a different setting which is just present in Kernel/Config/Default.pm and can just be changed in Kernel/Config.pm. This is the template for the actual index:

`    $Self->{'Elasticsearch::IndexSettings'} = {
        index => {
            number_of_shards   => '[% Data.NS | uri %]',
            number_of_replicas => '[% Data.NR | uri %]',
        },
        'index.mapping.total_fields.limit' => '[% Data.FieldsLimit | uri %]',
    };`

I would suggest to rename the template to Elasticsearch::IndexTemplate###Default. People should then be able to also put specific templates as Elasticsearch::IndexTemplate###Ticket, or Customer,... to Kernel/Config.pm. (This should be noted in a comment in Default.pm.)

svenoe replied:
Ah, sorry, the formatting here is bad. Please have a look for Elasticsearch::IndexSettings in Kernel/Config/Default.pm.

svenoe commented on Kernel/System/Console/Command/Maint/Elasticsearch/Migration.pm:
```diff
@@ -734,7 +744,12 @@ sub _ExpandTemplate {
+        }
+        else {
+            return $Node;
+        }
```
This whole block is likely not right, just because it is connected to my former comment.

kess-net replied:
Did you test this and experienced a wrong behavior - which I would like to see -  or do you just guess?

kess-net committed:
ssue #961 - PR 1165 be strict here

kess-net committed:
issue #961,  PR 1165 the installer now users FieldsLimit setting

kess-net committed:
Code style

use the logical defined or operator for better readability

kess-net committed:
issue #961 Installer now respects customizable index settings

Template and settings are predefined in Kernel/Config/Defaults.pm for
the indices Customer, CustomerUser and Ticket.

kess-net committed:
issue #961 fix variable name

kess-net committed:
issue #961 enablle index setting templates

Maint::Elasticsearch::Migration now uses Index templates the same way as
the installer

kess-net committed:
make CodePolicy.pl happy

kess-net committed:
issue #961 adapt test to latest code changes
